### PR TITLE
refactor: [M3-6063] - Refactor components to use TypeToConfirmDialog component

### PR DESCRIPTION
**Pull request by carrillo-erik**
## Description 📝
Refactor components to standardize type-to-confirm implementation using the `TypeToConfirmDialog` component.

## Major Changes 🔄
- Introduced the following new `types` in the `EntityInfo` interface: `Database`, `Kubernetes`, `AccountSetting`.
- Introduced the `subType` property in the `EntityInfo` interface: `Backup`, `Cluster`, `ObjectStorage`, `CloseAccount`.

## How to test 🧪
1. Verify type-to-confirm functionality across the application.
2. Check for any visual regressesions.

### CODE REVIEW

**Comment by bnussman** [comment]
Please add a chageset to your PR 🥺🚨

**Comment by bnussman** [comment]
Please add a chageset to your PR 🥺🚨

**Comment by abailly-akamai** [comment]
@carrillo-erik not sure what you are trying to do with passing `{...rest}` to `TypeToConfirmDialog`? Maybe start by deleting that and work your way from there?

Screenshot 2023-05-30 at 1 47 03 PM

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 1]
```diff
@@ -84,27 +131,31 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
...
-        value={confirmText}
-        typographyStyle={typographyStyle}
-        data-testid={'dialog-confirm-text-input'}
-        expand
-        onChange={(input) => {
-          setConfirmText(input);
-        }}
-        visible={preferences?.type_to_confirm}
-      />
+      <div style={{ display: 'flex', flexDirection: 'column', order: 2 }}>
```
Comment (bnussman-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:134]: I don't think this div is doing much. I think it maybe can removed. If we need it, I think we should use a `<Stack />`.

```suggestion
```

File: packages/manager/src/features/Account/EnableObjectStorage.tsx
Comment (bnussman-akamai) [inline packages/manager/src/features/Account/EnableObjectStorage.tsx:1]: This dialog has the wrong copy and also the `entity.type` saying "AccountSetting" needs to be improved.
Screenshot 2023-06-05 at 10 30 31 AM

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 2]
```diff
@@ -56,9 +64,49 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
     }
   }, [open]);
 
+  const getPrimaryBtnText = (entity: EntityInfo) => {
+    if (entity.type === 'Volume' && title.startsWith('Detach')) return 'Detach';
```
Comment (abailly-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:68]: isn't `title` an optional prop? seems odd that TS isn't complaining here

Comment (abailly-akamai) [reply packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:68]:
```suggestion
    if (entity.type === 'Volume' && title?.startsWith('Detach')) return 'Detach';
```

File: packages/manager/src/features/Account/CloseAccountDialog.tsx [hunk 3]
```diff
@@ -21,6 +18,9 @@ interface Props {
 const useStyles = makeStyles()((theme: Theme) => ({
   dontgo: {
     marginTop: theme.spacing(2),
+    display: 'flex',
+    flexDirection: 'column',
+    order: 3,
```
Comment (abailly-akamai) [inline packages/manager/src/features/Account/CloseAccountDialog.tsx:23]: I believe you don't need `display` or `flexDirection` to declare an `order` with a grid or flex parent

File: packages/manager/src/features/Account/CloseAccountDialog.tsx [hunk 4]
```diff
@@ -98,85 +97,58 @@ const CloseAccountDialog = ({ closeDialog, open }: Props) => {
...
+        style={{ display: 'flex', flexDirection: 'column', order: 1 }}
+      >
+        <Typography
+          style={{
+            fontSize: '0.875rem',
+            display: 'flex',
+            flexDirection: 'column',
+            order: 3,
+          }}
+        >
```
Comment (abailly-akamai) [inline packages/manager/src/features/Account/CloseAccountDialog.tsx:126]: same here, no need for flex and we should use a styled component

File: packages/manager/src/features/Account/CloseAccountDialog.tsx [hunk 5]
```diff
@@ -98,85 +97,58 @@ const CloseAccountDialog = ({ closeDialog, open }: Props) => {
...
-      <Button
-        buttonType="primary"
-        onClick={onSubmit}
-        disabled={disabled}
-        loading={isCanceling}
-      >
-        Close Account
-      </Button>
-    </ActionsPanel>
+      <div style={{ display: 'flex', flexDirection: 'column', order: 4 }}>
```
Comment (abailly-akamai) [inline packages/manager/src/features/Account/CloseAccountDialog.tsx:139]: same

File: packages/manager/src/features/Account/EnableObjectStorage.tsx [hunk 6]
```diff
@@ -134,35 +104,31 @@ export const EnableObjectStorage = (props: Props) => {
...
-        title="Cancel Object Storage"
-        actions={actions}
+        loading={isLoading}
+        typographyStyle={{ marginTop: '20px' }}
+        onClose={handleClose}
+        onClick={handleSubmit}
       >
+        {error ? <Notice error text={error} /> : null}
         <Notice warning>
           <Typography style={{ fontSize: '0.875rem' }}>
```
Comment (abailly-akamai) [inline packages/manager/src/features/Account/EnableObjectStorage.tsx:122]: let's get rid of style in favor of sx when possible

File: packages/manager/src/features/Account/CloseAccountDialog.tsx (hunk 4 above)
Comment (abailly-akamai) [reply packages/manager/src/features/Account/CloseAccountDialog.tsx:126]: also you already have an order 3 on the `dontgo` className - wondering what this does

**Review by abailly-akamai — COMMENTED** [review]
Nice! needs some cleanup (will review in the browser when done)

Also, don't hesitate to add a comment why you are using a flex order because of the renderProps

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 7]
```diff
@@ -56,9 +64,49 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
...
+        return 'Confirm Cancellation';
+      case 'CloseAccount':
+        return 'Close Account';
+      default:
+        return 'Delete';
+    }
+  };
+
+  const getLabelText = (entity: EntityInfo) => {
+    if (entity.type === 'AccountSetting' && entity.subType === 'CloseAccount') {
```
Comment (dwiley-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:85]: Are you seeing this linter complaint also?

Screenshot 2023-06-05 at 4 59 29 PM

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 8]
```diff
@@ -84,27 +131,31 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
...
-        onChange={(input) => {
-          setConfirmText(input);
-        }}
-        visible={preferences?.type_to_confirm}
-      />
+      <div style={{ display: 'flex', flexDirection: 'column', order: 2 }}>
+        <TypeToConfirm
+          hideInstructions={entity.subType === 'CloseAccount'}
+          label={getLabelText(entity)}
+          confirmationText={
```
Comment (dwiley-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:138]: Would there be a scenario under which `confirmationText` would be defined but shouldn't be rendered?

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 9]
```diff
@@ -56,9 +64,49 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
...
+  const getLabelText = (entity: EntityInfo) => {
+    if (entity.type === 'AccountSetting' && entity.subType === 'CloseAccount') {
+      return `Please enter your username (${entity.label}) to confirm.`;
+    }
+    const ttcLabelText =
+      entity.type === 'Bucket' || entity.subType === 'Cluster'
+        ? 'Name'
+        : 'Label';
+    switch (entity.subType) {
+      case 'ObjectStorage':
```
Comment (dwiley-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:93]: The distinction between `Bucket` and `ObjectStorage` here wasn't clear to me on the first pass or two -- had to check to ascertain that the latter was referring to canceling OBJ

File: packages/manager/src/features/Databases/DatabaseDetail/DatabaseSettings/DatabaseSettingsDeleteClusterDialog.tsx [hunk 10]
```diff
@@ -82,35 +46,23 @@ export const DatabaseSettingsDeleteClusterDialog: React.FC<Props> = (props) => {
...
-        data-testid={'dialog-confirm-text-input'}
-        label="Cluster Name"
-        onChange={(input) => setConfirmText(input)}
-        expand
-        value={confirmText}
-        confirmationText={
-          <span>
-            To confirm deletion, type the name of the database cluster (
-            <b>{databaseLabel}</b>) in the field below:
-          </span>
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Databases/DatabaseDetail/DatabaseSettings/DatabaseSettingsDeleteClusterDialog.tsx:109]: the removal of this results in some slightly different copy now with a capitalized entity name -- has UX weighed in?

Screenshot 2023-06-05 at 5 16 21 PM

File: packages/manager/src/features/Kubernetes/KubernetesClusterDetail/DeleteKubernetesClusterDialog.tsx [hunk 11]
```diff
@@ -51,37 +42,16 @@ export const DeleteKubernetesClusterDialog = (props: Props) => {
...
-      </Button>
-    </ActionsPanel>
-  );
-
   return (
-    <ConfirmationDialog
-      open={open}
+    <TypeToConfirmDialog
       title={`Delete Cluster ${clusterLabel}`}
+      entity={{ type: 'Kubernetes', subType: 'Cluster', label: clusterLabel }}
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Kubernetes/KubernetesClusterDetail/DeleteKubernetesClusterDialog.tsx:48]: leads to "the Kubernetes" with the current logic in `TypeToConfirmDialog.tsx`

Screenshot 2023-06-05 at 5 18 08 PM

File: packages/manager/src/features/Account/CloseAccountDialog.tsx
Comment (dwiley-akamai) [inline packages/manager/src/features/Account/CloseAccountDialog.tsx:1]: "Warning" is on its own line now compared to production:

Screenshot 2023-06-05 at 1 15 49 PM

File: packages/manager/src/features/Databases/DatabaseDetail/DatabaseBackups/RestoreFromBackupDialog.tsx
Comment (dwiley-akamai) [inline packages/manager/src/features/Databases/DatabaseDetail/DatabaseBackups/RestoreFromBackupDialog.tsx:1]: Looks like the copy got mixed up with database deletion:

Screenshot 2023-06-05 at 12 58 11 PM

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx [hunk 12]
```diff
@@ -56,9 +64,87 @@ export const TypeToConfirmDialog = (props: CombinedProps) => {
...
+      return '';
+    }
+
+    let ttcAction = '';
+    if (entity.subType === 'Backup') {
+      ttcAction = 'restoration';
+    } else if (entity.subType === 'ObjectStorage') {
+      ttcAction = 'cancellation';
+    } else {
+      ttcAction = 'deletion';
```
Comment (abailly-akamai) [inline packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx:117]: those would be much better as switch statements

File: packages/manager/src/features/Account/CloseAccountDialog.tsx
Comment (carrillo-erik) [reply packages/manager/src/features/Account/CloseAccountDialog.tsx:1]: I have resolved this issue with my latest changes.

File: packages/manager/src/features/Kubernetes/KubernetesClusterDetail/DeleteKubernetesClusterDialog.tsx (hunk 11 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Kubernetes/KubernetesClusterDetail/DeleteKubernetesClusterDialog.tsx:41]: I've updated the logic to address this issue.

File: packages/manager/src/features/Account/EnableObjectStorage.tsx
Comment (carrillo-erik) [reply packages/manager/src/features/Account/EnableObjectStorage.tsx:1]: I think I must have inadvertently copied and pasted code from a different component here and it slipped by. I have fixed this issue.

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx (hunk 1 above)
Comment (carrillo-erik) [reply packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx]: After addressing other comments, I realized this was not necessary. It's been removed.

File: packages/manager/src/features/Account/CloseAccountDialog.tsx (hunk 3 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Account/CloseAccountDialog.tsx:21]: You are correct. I think I had my components mixed up and left these properties on accident.

File: packages/manager/src/features/Account/CloseAccountDialog.tsx (hunk 4 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Account/CloseAccountDialog.tsx]: Thanks for the second pair of eyes on this. I've made the necessary changes.

File: packages/manager/src/features/Account/EnableObjectStorage.tsx (hunk 6 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Account/EnableObjectStorage.tsx:121]: Done ✅

File: packages/manager/src/features/Databases/DatabaseDetail/DatabaseSettings/DatabaseSettingsDeleteClusterDialog.tsx (hunk 10 above)
Comment (carrillo-erik) [reply packages/manager/src/features/Databases/DatabaseDetail/DatabaseSettings/DatabaseSettingsDeleteClusterDialog.tsx:109]: I've changed the logic to keep it consistent to the previous implementation.

File: packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx (hunk 12 above)
Comment (carrillo-erik) [reply packages/manager/src/components/TypeToConfirmDialog/TypeToConfirmDialog.tsx]: Took your suggestion and made the change.

**Review by dwiley-akamai — COMMENTED** [review]
I went through the app on this branch and compared and to prod and noticed:

- There's a bit of extra padding below the Warning notice in the `Close Account` modal and below the Warning notice in the `Delete OBJ Bucket` modal
- There's a lot of extra padding below the input field in the Linode Resize & Rebuild modals
- Still seeing "To confirm deletion..." (instead of "To confirm cancellation...") in the `Cancel Object Storage` modal

File: packages/manager/src/features/Databases/DatabaseDetail/DatabaseBackups/RestoreFromBackupDialog.tsx [hunk 13]
```diff
@@ -49,40 +43,16 @@ export const RestoreFromBackupDialog: React.FC<Props> = (props) => {
...
-      {...rest}
+    <TypeToConfirmDialog
       title={`Restore from Backup ${formatDate(backup.created, {
         timezone: profile?.timezone,
       })}`}
+      entity={{ type: 'Database', subType: 'Backup', label: database.label }}
       open={open}
       onClose={onClose}
-      actions={actions}
+      onClick={() => handleRestoreDatabase}
```
[the remark this answers is not in the document]
Comment (carrillo-erik) [reply packages/manager/src/features/Databases/DatabaseDetail/DatabaseBackups/RestoreFromBackupDialog.tsx]: This was the one issue I was having the TS and the `onClick` event when I started refactoring the component. I've made several changes from when I first started working on this ticket. I went back and revert the changes to this event handler and TS doesn't complaint anymore and the functionality works.

**Comment by carrillo-erik** [comment]
I've addressed these concerns with my latest changes.

File: packages/manager/src/features/Account/CloseAccountDialog.tsx
Comment (hkhalil-akamai) [inline packages/manager/src/features/Account/CloseAccountDialog.tsx:1]: Noticing a regression in this component as the "Close Account" button is never enabled, even when the correct username is entered in the confirmation field. Looks like the `inputtedUsername` state is never being updated.

**Review by hkhalil-akamai — CHANGES_REQUESTED** [review]

**Review by bnussman-akamai — CHANGES_REQUESTED** [review]
Can you checkout the PR and run `yarn`, then push the changes. The `yarn.lock` is out of date

**Comment by jaalah-akamai** [comment]
@carrillo-erik There has come together really nice. I think this component will benefit a ton from a storybook page. See https://github.com/linode/manager/tree/develop/packages/manager/src/components/Button as to how we did the JSDocs and stories file.

**Comment by carrillo-erik** [comment]
@bnussman-akamai I had pushed some changes to the `yarn.lock` file and the `cachedData` files. John asked me to remove those changes from the PR. I reverted the commit and reintroduced the changes to fix the pertaining issue.

**Review by hkhalil-akamai — APPROVED** [review]
Functionality of all improved components looks good. Approved pending suggestions from @jaalah-akamai @bnussman-akamai.

Nice work and thanks for bearing with the review changes!

**Review by cpathipa — APPROVED** [review]

**Comment by bnussman-akamai** [comment]
Sounds good @carrillo-erik, still before you merge, just run `yarn` and push those changes. If we don't we'll get dependabot alerts

**Comment by carrillo-erik** [comment]
> Can you checkout the PR and run `yarn`, then push the changes. The `yarn.lock` is out of date

I had some merge conflicts with the `yarn.lock` file after pulling the latest. I believe this resolves the issue you brought up. Let me know, thanks.

**Pull request merged**